### PR TITLE
Add delete method to sftp client

### PR DESCRIPTION
**Pull request by Diogenesoftoronto**

### CODE REVIEW

File: internal/sftp/client.go [hunk 1]
```diff
@@ -14,4 +14,7 @@ type Client interface {
 	// Upload transfers data from the provided source reader to a specified
 	// destination on the SFTP server.
 	Upload(ctx context.Context, src io.Reader, dest string) (bytes int64, err error)
+	// Delete removes data from the provided dest to a specified
```
Comment (djjuhasz) [inline internal/sftp/client.go]: Suggested comment: 
```
// Delete deletes the dest file from the SFTP server.
```

File: internal/sftp/client.go [hunk 2]
```diff
@@ -14,4 +14,7 @@ type Client interface {
...
+	// destination on the SFTP server.
+	Delete(ctx context.Context, dest string) (bytes int64, err error)
```
Comment (djjuhasz) [inline internal/sftp/client.go]: I can't think of any reason we'd need to know the size of the file after deleting it - I think the `bytes` return value should be removed.

File: internal/sftp/goclient.go [hunk 3]
```diff
@@ -57,6 +57,32 @@ func (c *GoClient) Upload(ctx context.Context, src io.Reader, dest string) (int6
 	return bytes, nil
 }
 
+// Delete removes the data from dest and returns the
```
Comment (djjuhasz) [inline internal/sftp/goclient.go]: See above suggestion for updating this comment. Also the comment text should be re-wrapped at 80 cols.

File: internal/sftp/goclient.go [hunk 4]
```diff
@@ -57,6 +57,32 @@ func (c *GoClient) Upload(ctx context.Context, src io.Reader, dest string) (int6
...
+//
+// Delete is not thread safe.
+func (c *GoClient) Delete(ctx context.Context, dest string) (int64, error) {
+	if err := c.dial(); err != nil {
+		return 0, err
+	}
+	defer c.close()
+
+	// Get file information to find out its size
+	fileInfo, err := c.sftp.Lstat(dest)
```
Comment (djjuhasz) [inline internal/sftp/goclient.go]: I think even without returning `bytes` the lstat is a good check that the file exists before deleting it, so we can provide a better error message.

File: internal/sftp/goclient.go [hunk 5]
```diff
@@ -57,6 +57,32 @@ func (c *GoClient) Upload(ctx context.Context, src io.Reader, dest string) (int6
...
+	if err != nil {
+		return 0, fmt.Errorf("SFTP: unable to get file info for %q: %v", dest, err)
```
Comment (djjuhasz) [inline internal/sftp/goclient.go]: Suggested error message: `fmt.Errorf("SFTP: remote file doesn't exist or is unreadable: %s", dest)`

File: internal/sftp/goclient_test.go [hunk 6]
```diff
@@ -287,4 +286,60 @@ func TestGoClient(t *testing.T) {
 			assert.Assert(t, tfs.Equal(dest.Path(), tfs.Expected(t, tc.want.Paths...)))
 		})
 	}
+
+	type deleteTest struct {
```
Comment (djjuhasz) [inline internal/sftp/goclient_test.go]: I think it's better to move this to it's own test, e.g.

```
func TestDelete(t *testing.T) {
```

And rename the first test `TestUpload` (or similar).

File: internal/sftp/goclient_test.go [hunk 7]
```diff
@@ -287,4 +286,60 @@ func TestGoClient(t *testing.T) {
...
+	} {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			sftpc := sftp.NewGoClient(tc.cfg)
+			src := strings.NewReader("Testing 1-2-3")
+			dest := tfs.NewDir(t, "sftp_test")
+
+			sftpc.Upload(context.Background(), src, dest.Join("test.txt"))
```
Comment (djjuhasz) [inline internal/sftp/goclient_test.go]: This test will fail if the `Upload()` method fails, which isn't ideal.  Because the SFTP server is using the same filesystem as the client, you can create a file directly in `dest` then call `sftpc.Delete()` to delete it.

**Review by djjuhasz — CHANGES_REQUESTED** [review]
@Diogenesoftoronto overall this looks good, but I've requested a few changes.

File: internal/sftp/goclient.go (hunk 4 above)
Comment (sevein) [reply internal/sftp/goclient.go]: An alternative where we do it in one go to save us the extra round-trip:

```go
func (c *GoClient) Delete(ctx context.Context, dest string) error {
	if err := c.dial(); err != nil {
		return fmt.Errorf("SFTP: unable to dial: %v", err)
	}
	defer c.close()

	err := c.sftp.Remove(dest)
	if err != nil {
		// Wrap only non-exist and permission errors to avoid leaking.
		if os.IsNotExist(err) || os.IsPermission(err) {
			return fmt.Errorf("SFTP: unable to remove file %q: %w", dest, err)
		}
		return fmt.Errorf("SFTP: unable to remove file %q: %v", dest, err)
	}

	return nil
}
```

Notice that I'm only using `%w` when I want to expose `os.ErrNotExist` or `os.ErrPermission`, but using `%v` to avoid other errors that could lead to implementation leaking.

File: internal/sftp/goclient.go [hunk 8]
```diff
@@ -57,6 +57,32 @@ func (c *GoClient) Upload(ctx context.Context, src io.Reader, dest string) (int6
 	return bytes, nil
 }
 
+// Delete removes the data from dest and returns the
+// number of bytes removed. A new SFTP connection is opened before removing the file, and
+// closed when the delete is complete or cancelled.
+//
+// Delete is not thread safe.
+func (c *GoClient) Delete(ctx context.Context, dest string) (int64, error) {
```
Comment (sevein) [inline internal/sftp/goclient.go]: I'm realizing that we can't use ctx for cancellation, but I suggest to leave it in case we find a way to use it in the future. I think that both the ssh and sftp packages that we're using predate the context package so they were not designed to accept a context. One alternative that we could consider is to pass the context to the dial method and create our own dialer with [net.DialContext](https://pkg.go.dev/net#Dialer.DialContext), but I don't see it as a blocker.

Comment (sevein) [reply internal/sftp/goclient.go]: rclone is doing it [here](https://github.com/rclone/rclone/blob/e1b0417c286bdcf5e866d70d51cb75be179c25e6/backend/sftp/ssh_internal.go#L26-L48), but it looks complicated because they support edge cases like proxy settings, etc. It also looks like the ssh package is going to have a DialContext sometime soon: https://go-review.googlesource.com/c/crypto/+/504735.

Comment (djjuhasz) [reply internal/sftp/goclient.go]: Huh, interesting @sevein.  I thought ctx would be unnecessary for delete because the actual `sftp.Remove()` operation should be very fast, AFAIK.  But now I realize that dialing the server could take a while if there are communications issues, so I agree ctx is still worthwhile for future use.

File: internal/sftp/goclient.go (hunk 4 above)
Comment (djjuhasz) [reply internal/sftp/goclient.go]: I like this alternative, but I would make a small tweak to the non-exist/permission error message:
```
return fmt.Errorf("SFTP: unable to remove file: %w", err)
```
I find that most (maybe all?) of the filesystem error messages include the name of the file, so adding it in the wrapper causes duplication.  That said, it's easy to test for `os.IsNotExist(err)` in the unit tests so it's probably worth checking what the actual error message looks like.

Comment (sevein) [reply internal/sftp/goclient.go]: You're right. I had a quick look at the code and it looks like it only returns a sentinel error but it's better to test it.

File: internal/sftp/goclient.go (hunk 8 above)
Comment (Diogenesoftoronto) [reply internal/sftp/goclient.go]: One thing I have considered is doing something similar to what contextio does, or wwe can do @sevein suggestion. I think including context on the actual client struct makes sense and then we can context is equal for operations that normally do not have a context like in contextio

File: internal/sftp/goclient_test.go (hunk 7 above)
Comment (Diogenesoftoronto) [reply internal/sftp/goclient_test.go]: Yes this is a better idea.

File: internal/sftp/goclient_test.go (hunk 6 above)
Comment (Diogenesoftoronto) [reply internal/sftp/goclient_test.go]: Yes I was considering this. I will go through with this.

File: internal/sftp/client.go (hunk 1 above)
Comment (Diogenesoftoronto) [reply internal/sftp/client.go]: destination can be a directory.

File: internal/sftp/goclient_test.go [hunk 9]
```diff
@@ -288,3 +287,75 @@ func TestGoClient(t *testing.T) {
...
+	host, port, err := net.SplitHostPort(serverAddress)
+	if err != nil {
+		t.Fatalf("Bad server address: %s", serverAddress)
+	}
+
+	_ = startSFTPServer(t, serverAddress)
+
+	// Start a listener on an open port and use the address to test a bad SFTP
+	// server address.
+	listener, err := net.Listen("tcp", "127.0.0.1:0")
```
Comment (djjuhasz) [inline internal/sftp/goclient_test.go]: `listener` is unused.

File: internal/sftp/goclient_test.go [hunk 10]
```diff
@@ -288,3 +287,75 @@ func TestGoClient(t *testing.T) {
...
+			name: "Deletes a file",
+			cfg: sftp.Config{
+				Host:           host,
+				Port:           port,
+				KnownHostsFile: "./testdata/known_hosts",
+				PrivateKey: sftp.PrivateKey{
+					Path: "./testdata/clientkeys/test_ed25519",
+				},
+			},
+			bytes:    13,
```
Comment (djjuhasz) [inline internal/sftp/goclient_test.go]: `bytes` is unused.

File: internal/sftp/goclient_test.go [hunk 11]
```diff
@@ -288,3 +287,75 @@ func TestGoClient(t *testing.T) {
...
+			wantErr:  "SFTP: unable to get file info for",
+		},
+	} {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			sftpc := sftp.NewGoClient(tc.cfg)
+			dest := tfs.NewDir(t, "sftp_test")
+			pre := dest.Join("test.txt")
```
Comment (djjuhasz) [inline internal/sftp/goclient_test.go]: This can be combined with the `tfs.NewDir()` method:

```
dest := tfs.NewDir(t, "sftp_test", tfs.WithFile("test.txt", ""))
```

File: internal/sftp/goclient_test.go [hunk 12]
```diff
@@ -227,8 +227,7 @@ func TestGoClient(t *testing.T) {
 			},
 			wantErr: fmt.Sprintf(
 				"SSH: connect: dial tcp %s:%s: connect: connection refused",
-				badHost, badPort,
-			),
+				badHost, badPort),
```
Comment (djjuhasz) [inline internal/sftp/goclient_test.go]: Why is this change necessary?

File: internal/sftp/goclient.go [hunk 13]
```diff
@@ -57,6 +57,26 @@ func (c *GoClient) Upload(ctx context.Context, src io.Reader, dest string) (int6
...
+		return fmt.Errorf("SFTP: unable to dial: %w", err)
+	}
+	defer c.close()
+
+	// Delete the file
+	if err := c.sftp.Remove(dest); err != nil {
+		if os.IsNotExist(err) {
+			return fmt.Errorf("SFTP: file does not exist: %w", err)
+		} else if os.IsPermission(err) {
+			return fmt.Errorf("SFTP: insufficient conditions to delete file: %w", err)
```
Comment (djjuhasz) [inline internal/sftp/goclient.go]: I think "permissions" is clearer than "conditions".  Also, as you pointed out, dest could be a dir.

Suggested:
```
return fmt.Errorf("SFTP: insufficient permissions to delete: %w", err)
```

**Review by djjuhasz — CHANGES_REQUESTED** [review]
@Diogenesoftoronto a few more change requests.

File: internal/sftp/goclient.go (hunk 13 above)
Comment (Diogenesoftoronto) [reply internal/sftp/goclient.go]: good catch and good point

File: internal/sftp/goclient_test.go (hunk 12 above)
Comment (Diogenesoftoronto) [reply internal/sftp/goclient_test.go]: its a result of lint i think

File: internal/sftp/goclient_test.go (hunk 11 above)
Comment (Diogenesoftoronto) [reply internal/sftp/goclient_test.go]: it can be but I think the way it is clearer

Comment (Diogenesoftoronto) [reply internal/sftp/goclient_test.go]: actually nope i agree with you now. just had to see it.

File: internal/sftp/goclient_test.go (hunk 12 above)
Comment (djjuhasz) [reply internal/sftp/goclient_test.go]: I just ran the linter on main and it didn't change this. :confused:

**Review by djjuhasz — APPROVED** [review]
Looks good. Don't forget to squash the commits.
